matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -161,84 +166,201 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
 
-        for (GPXEntry gpxEntry : gpxList) {
-            if (prevTimeStep == null
-                    || distanceCalc.calcDist(
-                            prevTimeStep.observation.getLat(), prevTimeStep.observation.getLon(),
-                            gpxEntry.getLat(), gpxEntry.getLon()) > 2 * measurementErrorSigma
-                    // always include last point
-                    || indexGPX == gpxList.size() - 1) {
-                final List<QueryResult> queryResults = locationIndex.findNClosest(
-                        gpxEntry.lat, gpxEntry.lon,
-                        edgeFilter, measurementErrorSigma);
-                outAllCandidates.addAll(queryResults);
-                final List<GPXExtension> candidates = new ArrayList<>();
-                for (QueryResult candidate : queryResults) {
-                    candidates.add(new GPXExtension(gpxEntry, candidate, indexGPX));
+        int n = filteredGPXEntries.size();
+        assert queriesPerEntry.size() == n;
+        for (int i = 0; i < n; i++) {
+            
+            GPXEntry gpxEntry = filteredGPXEntries.get(i);
+            List<QueryResult> queryResults = queriesPerEntry.get(i);
+
+            /*
+             * If the GPS measurement is mapped to a virtual node then we need to create a
+             * candidate for each direction (see also #51).
+             */
+            List<GPXExtension> candidates = new ArrayList<>();
+            for (QueryResult qr: queryResults) {
+                int closestNode = qr.getClosestNode();
+                if (queryGraph.isVirtualNode(closestNode)) {
+                    // get virtual edges:
+                    List<VirtualEdgeIteratorState> virtualEdges = new ArrayList<>();
+                    EdgeIterator iter = queryGraph.createEdgeExplorer().setBaseNode(closestNode);
+                    while (iter.next()) {
+                        if (queryGraph.isVirtualEdge(iter.getEdge())) {
+                            virtualEdges.add((VirtualEdgeIteratorState)
+                                    queryGraph.getEdgeIteratorState(iter.getEdge(), iter.getAdjNode()));
+                        }
+                    }
+                    assert virtualEdges.size() == 2;
+
+                    // create a candidate for each: the candidate being the querypoint plus the
+                    // virtual edge to favour. Note that we favour the virtual edge by
+                    // *unfavoring* the rest, so we need to record these.
+                    VirtualEdgeIteratorState e1 = virtualEdges.get(0);
+                    VirtualEdgeIteratorState e2 = virtualEdges.get(1);
+                    for (int j = 0; j < 2; j++) {
+                        // get favored/unfavored edges:
+                        VirtualEdgeIteratorState incomingVirtualEdge = j == 0 ? e1 : e2;
+                        VirtualEdgeIteratorState outgoingVirtualEdge = j == 0 ? e2 : e1;
+                        // create candidate
```

Comment:
If we would create a candidate with incomingVirtualEdge==outgoingVirtualEdge then this candidate would actually allow to perform a U-turn without a penalty by going to and from the virtual node through the other virtual edge pair.